### PR TITLE
[Ready for Code Review] feature/spam_admin

### PR DESCRIPTION
**Pull request by himanshuo**
<h1>Purpose</h1>
This PR serves as the OSF connection to the Spam Assassin service created previously. It is a administration page for 'spam_admin' users to view a list of comments or projects that have been marked as spam. This serves to facilitate some human at COS in determining which things are spam or not. A notable issue this PR solves is that there is no conclusive way to determine what occurs when a comment is reported yet. This will fix this issue by allowing a human user to determine if the comment is actually spam.

<h1>Changes</h1>
When a user makes a new comment or edits an old one, it will be checked for spam by sending the comment to the spam assassin service. Comments will be marked as spam when the spam assassin service says so or when a user reports a comment as spam a given number of times (1 for now). For projects, as soon as a user tries to make his/her project public, we will send the project to the spam assassin service. If the service determines the project has spam elements, we mark it as spam. The user will still have full access to his/her project. The only difference is that the user will not be able to make the project public. The user will be alerted appropriately. 

There is a separate admin page ('localhost:5000/spam_admin/') that allows a human COS people to easily view information about each comment or project that has been marked as possibly spam (as described above). We can now read through the comment and actually determine if it is spam or not. If we mark the comment as spam, it will delete the comment. In addition, it will also train the spam assassin service to make it smarter for the future. If we mark the comment as ham, then it will be treated as a normal non-spam comment in the future. The idea is similar for projects. The only difference is that we will not delete a project if it is marked as spam. Notably, if a comment or project was already conclusively determined to be ham or spam by a spam_admin, we will not check once again as it likely not going to change.

To become a spam_admin, a user (referring to a OSF class User ) has to have a spam_admin boolean field be equal to true. Non-spam_admin's can't access this page. Currently, a user that has email spam_admin@cos.com and full name spam_admin will automatically become a spam_admin.

There are also tests testing both the views of the spam_admin page and the functional changes in handling comments and making projects public. 

<h1>Side Effects</h1>
There can be a notable slowdown in making a project public if it is large enough. This is due to checking for whether it is spam or not.

### CODE REVIEW

File: tests/test_spam_admin.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+    @httpretty.activate
+    def test_comment_added_is_spam(self):
+        comment = self._add_comment(
+            self.project, content = self.GTUBE, auth=self.project.creator.auth,
+        )
+        self.project.reload()
+
+
+
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: Are the multi-line spaces within a method intentional?

File: tests/test_spam_admin.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+        assert_equal(len(self.project.commented), 1)
+
+        if self.spam_assassin_active:
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: Are you testing with Spam Assassin active, not, or both sequentially? This case and similar should be split into two test cases, since it could test two different instances .

File: tests/test_spam_admin.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+            assert_false(train_spam(comment, is_spam=False))
+
+    @httpretty.activate
+    def test_auto_mark_spam_if_flagged_enough_times(self):
+        comment = self._add_comment(
+            self.project, auth=self.project.creator.auth,
+        )
+
+
+        for i in range(Comment.NUM_FLAGS_FOR_SPAM):
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: Looking at model.py NUM_FLAGS_FOR_SPAM is set to 1. Is this the correct variable initialization? If it is I would consider a different number for testing purposes, as 1 is an edge-case in itself.

File: tests/test_spam_admin.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        assert_equal(comment.spam_status , Comment.POSSIBLE_SPAM)
+
+    @httpretty.activate
+    def test_dont_auto_mark_spam_if_not_flagged_enough_times(self):
+        comment = self._add_comment(
+            self.project, auth=self.project.creator.auth,
+        )
+
+
+        for i in range(Comment.NUM_FLAGS_FOR_SPAM-1):
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: See previous comment. Where 1 is an edge-case, 0 means that this code won't even run once meaning that the post won't happen and the code that it is testing won't be run a single time. Using a larger number for test cases will ensure that the code is run and it won't trip something in the first run.

File: tests/test_spam_admin.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+            )
+
+        ret=self.app.get(
+            '/api/v1/spam_admin/list_comments/3/',
+            auth=self.spam_admin.auth
+        )
+
+
+        if self.spam_assassin_active:
+            #todo: stop comments from earlier tests from saving
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: Can you get the number of current comments in the list at the beginning of the test and assert the number of added comments from the difference?

File: website/project/views/node.py [hunk 6]
```diff
@@ -38,7 +38,7 @@
 from website.profile import utils
 from website.project import new_folder
 from website.util.sanitize import strip_html
-
+from website.spam_admin.utils import _project_is_spam
```
Comment (Ghalko) [inline website/project/views/node.py]: Not sure an '_' should be the first character on something that needs to be imported.

File: website/spam_admin/utils.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        if json.loads(resp.text).get('decision', '') == "SPAM":
+            return True
+        return False
+    except:
+        return False
+
+def train_spam_project(project, is_spam):
+    if not settings.SPAM_ASSASSIN:
+            return False
+    try:
```
Comment (Ghalko) [inline website/spam_admin/utils.py]: What exception will be thrown here? Why is a try block being used?

File: website/spam_admin/views.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+@must_be_logged_in
+@must_be_spam_admin
+def list_comment_page(**kwargs):
+    """
+    make a list of comments that are marked as possibleSpam
+    """
+    if not spam_assassin_active:
+        return {
+            'comments': list([]),
```
Comment (Ghalko) [inline website/spam_admin/views.py]: How will seeing an empty list of comment spam tell the spam_admin that the spam assassin server is not running?

File: website/templates/spam_admin_projects.mako [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                </div>
+
+
+                <div class="panel panel-default">
+                  <!-- Default panel contents -->
+                  <div class="panel-heading">Components</div>
+
+                  <!-- List group -->
+                  <ul class="list-group">
+                      <!-- ko foreach: project.components -->
```
Comment (Ghalko) [inline website/templates/spam_admin_projects.mako]: Why does this need ko comment syntax? The rest of the PR doesn't rely on this method, including line 57 of this file, for a specific foreach case.

File: tests/test_spam_admin.py [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+    @httpretty.activate
+    def test_comment_added_is_spam(self):
+        comment = self._add_comment(
+            self.project, content = self.GTUBE, auth=self.project.creator.auth,
+        )
+        self.project.reload()
+
+
+
+        assert_equal(len(self.project.commented), 1)
```
Comment (Ghalko) [inline tests/test_spam_admin.py]: Are the multi-line spaces within a method intentional?

**Comment by Ghalko** [comment]
First pass done. This is a learning experience for me, so please don't be afraid to discuss the questions/suggestions with me.

File: tests/test_spam_admin.py (hunk 10 above)
Comment (himanshuo) [inline tests/test_spam_admin.py]: yea, i like lots of spaces since it helps separate code. I'll remove them from the tests.

File: tests/test_spam_admin.py (hunk 3 above)
Comment (himanshuo) [inline tests/test_spam_admin.py]: good point. The idea was to have it set to 1 so I tested it with that as well. I will write more tests with higher values for NUM_FLAGS_FOR_SPAM. I'll also have tests with the edge case of NUM_FLAGS_FOR_SPAM=1.

File: website/spam_admin/utils.py (hunk 7 above)
Comment (himanshuo) [inline website/spam_admin/utils.py]: We do not want the OSF to be dependent on Spam Assassin whatsoever. Spam Assassin is only an accessory. Thus, this blanket try-catch statement serves to catch any and all errors that may occur due to calling the separate Spam Assassin service. The idea here is to have this entire method fail silently if anything goes wrong.

File: tests/test_spam_admin.py (hunk 5 above)
Comment (himanshuo) [inline tests/test_spam_admin.py]: ... smh. Yes. Wonderful idea.

File: tests/test_spam_admin.py (hunk 2 above)
Comment (himanshuo) [inline tests/test_spam_admin.py]: I am not sure how to do this. Spam Assassin could be active or inactive depending on the settings as defined in website.settings. These settings affect how many of the functions will work. I cannot change what the settings are from inside these tests. Therefore, I decided to just use an if-then to check what occurs in either case.

Comment (himanshuo) [inline tests/test_spam_admin.py]: Actually, just learned a lesson in python - I can actually import the settings directly (import website.settings) instead of getting a copy of the settings (using the from website.settings import ... syntax). This allows me to then actually manipulate the website settings sitewide. Using this, I've created new tests. I, now, test for when website.settings.SPAM_ASSASSIN = True and when website.settings.SPAM_ASSASSIN = False separately.

File: website/project/views/node.py (hunk 6 above)
Comment (himanshuo) [inline website/project/views/node.py]: When I first created it, this function was inside Node.py, thus it made sense to make it hidden. However, I eventually moved it to spam_admin utils to in order to centralize all my code. It is still not being used anywhere else so I thought making it anonymous is okay. However, I do see your point. I am importing it so its not technically hidden anymore. I will change it.

File: website/templates/spam_admin_projects.mako (hunk 9 above)
Comment (himanshuo) [inline website/templates/spam_admin_projects.mako]: In all previous usages of foreach, I repeated div's. In this case however, I wanted to repeat list items. The normal syntax ( data="foreach{data:project.components}" ) does not work for list items, thus this special comment syntax was required.

I could have use the ko foreach syntax with divs by repeating div's with a list item inside them, but that creates unnecessary div's.

File: website/spam_admin/views.py (hunk 8 above)
Comment (himanshuo) [inline website/spam_admin/views.py]: I think was too afraid of having the controller break and so had it just return an empty list if spam assassin is inactive. I had the view just be blank list in this case. Considering that spam admin should not be dependent on spam assassin, I have removed the entire if statement.

File: website/templates/spam_admin_projects.mako (hunk 9 above)
Comment (Ghalko) [inline website/templates/spam_admin_projects.mako]: Is this a known issue? It seems as though the ko docs show the data-bind syntax: http://knockoutjs.com/documentation/foreach-binding.html, http://knockoutjs.com/examples/collections.html

Is this because of how project.components is accessed?

Comment (himanshuo) [inline website/templates/spam_admin_projects.mako]: hmm, yea, you are right. I'm using the data-bind syntax right now and it seems to be working. I remember making a conscious decision to use the comment syntax earlier, but it seems to be working fine without it. In order to be consistent, I'll move forward with the data-bind syntax.

**Comment by himanshuo** [comment]
I believe I fixed the issues that were brought up.

File: website/spam_admin/decorators.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+from framework.auth.core import Auth
+def must_be_spam_admin(func):
+    """Require that user be spam_admin.
+    """
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+
+        kwargs['auth'] = Auth.from_kwargs(request.args.to_dict(), kwargs)
+        #todo: write separate script that creates new spam_admin users.
+        if kwargs['auth'].user.fullname == "spam_admin" and kwargs['auth'].user.emails[0] == "spam_admin@cos.com":
```
Comment (icereval) [inline website/spam_admin/decorators.py]: A system tag would be more appropriate than a special case for email.

**Comment by sloria** [comment]
Closing this for now. Development of the spam admin is continuing on the `feature/admin-app` branch.
